 ## Classical Mechanics - Week 6
 
 
 ### Last Week:
- Learned how to utilize 2D arrays in Python 
- Studied motion with quadratic drag with Euler's method
- Developed our computational skills in representing vectors with arrays

### This Week:
- Gain more practice with Python functions
- Learn and use the Trapezoidal Rule
- Further explore numerical methods

Do you smell that? It's midterm season, so let's take it easy this week, while getting some more familiarity with functions and numerical analysis. More specifically: we are going to see the power of numerical integration using the [**Trapezoidal Rule**.](https://en.wikipedia.org/wiki/Trapezoidal_rule)

In [1]:
# As usual, we will need packages
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

from IPython.display import Image
from IPython.core.display import HTML 

# [Numerical Integration](https://en.wikipedia.org/wiki/Numerical_integration)
What is numerical integration? To put it simply, it's a way of estimating the definite integral (area under a curve) of a function. Rather than finding the analytic form of the indefinite integral and evaluating at the endpoints, we will be using the function itself to calculate/estimate the definite integral. 

One of the simplest methods is known as the [Trapezoidal Rule.](https://en.wikipedia.org/wiki/Trapezoidal_rule)

The trapezoidal rule calculates the area under a curve, $f(x)$, from $a$ to $b$ by splitting it into a discrete number of sections, $N$, and approximating the function by a straight line within each section. As we learned before, the step size between points will be $h = \dfrac{b-a}{N}$. The following figure gives the basic idea used in the approximation (the step size is labeled $p$ in the figure).

![wikipidea](https://upload.wikimedia.org/wikipedia/commons/d/d1/Integration_num_trapezes_notation.svg)

The area of one of the trapezoids is $\dfrac{h}{2}\Bigl(f(x_i)+f(x_{i+1})\Bigr)$. Adding the contributions from each trapezoid gives our formula for the trapezoidal rule:

$\int_a^b f(x) \,dx \approx \dfrac{h}{2}\Sigma_{i=1}^N (f(x_{i-1})+f(x_i)) = \dfrac{h}{2}(f(x_0)+2f(x_1)+2f(x_2)+...+2f(x_{N-1})+f(x_N))$

Let's give this method a go!

Let's evaluate the integral of $f(x) = 4x^2 + 2x - 3$ from $x = 2$ to $x =10$. We will provide you with a routine that performs the Trapezoidal rule. Your job is to make the function we want to evaluate in the cell below. 

***Hint:*** Look at the Python Basics notebook we looked at together as a class if you need a refresher on how to make a function that returns equations. 

In [2]:
# First we must define a function we want to evaluate
def F(x):
    return 4*x**2+2*x-3# Insert the equation we want to integrate here

Now we must create a function that performs the Trapezoidal method. We will make ours take in a function (f), starting point (a), ending point (b), and number of steps we want to use (N). It will then return our approximated integration (I). Run the cell below to create our Trapezoidal method function, but read it to make sure you understand how we are creating this function.

In [3]:
# Creates our trapezoidal rule funciton
def Trapezoidal(f,a,b,N):

    h = (b-a)/N ## Calculates our step size, h
    I = f(a)+f(b) ## Adds f(a) and f(b) (end point contributions)
    
    for i in range(1,N): # Goes from 1 to N-1
        x=a+i*h  # calculate x_i
        I += 2*f(x) # performs summation of 2*f(x_i)
        
    return I*(h/2) #returns (h/2)*summation

Now in the cell below, use the Trapezoidal rule to evaluate $f(x)$ from $2$ to $10$ using 10 steps.

In [4]:
# Use our Trapezoidal function here
Trapezoidal(F, 2, 10, 10)

1398.08

This function is simple enough to perform an exact analytical integration.  Let's check this against our numerical calculation.  You may find the following cell useful to evaluate your exact expression of the integral.

In [5]:
def Fint(x):
    return 4/3*x**3+x**2-3*x

Fint(10)-Fint(2)

1394.6666666666665

# Q1.) What is the EXACT definite integral of $f(x) = 4x^2 + 2x - 3$ between $x=2$ and $x=10$? Round to the second decimal point. How does our approximation compare to the actual value?

Exact integral is 1394.67
Trapezoidal rule with N=10 gave 1398.08, which overshoots by about 0.24 percent.

# But why use numerical analysis instead of just doing the integration analytically?

A great question!

Our function $f(x)=4x^2+2x-3$ was pretty simple and easy to integrate analytically. However, this is usually not the case.  In fact, in many situations numerical integration is the only choice.

# Is our Trapezoidal method all we need for numerical integration?

Another great question!

Just like the Euler's method for solving differential equations, the Trapezoidal rule for numerical integration usually works fine given enough steps $N$, but it can be improved upon. These improved routines can allow us to get a better approximation with much less computer time, which can be crucial for multi-dimensional integrals.  There are
also methods designed to integrate badly behaved functions, where the Trapezoidal rule runs into problems. 

Try integrating the same function $f(x)=4x^2+2x-3$ using our Trapezoid routine with $N=100$ steps in the cell below in order to see how much better it does when we increase the number of steps by a factor of 10.

In [6]:
Trapezoidal(F, 2, 10, 100)

1394.7007999999996

## We will now use a library, SciPy, which contains many useful routines for scientific computing, to numerically evaluate this integral in the cell below.


Rather than always re-inventing the wheel, it is often easier to use packages, written by others, which include more advanced methods.

Run the cell below to see how the SciPy library can be used to solve a numerical integral.

In [16]:
# The following line imports a set of numerical integration routines from SciPy:
from scipy import integrate

# quad is a general purpose numerical integration routine in the subpackage scipy.integrate

integrate.quad(F,2,10)

(1394.6666666666667, 1.548391045010552e-11)

For the record, this routine assumes that the function can be complex, so it gives both a real and an imaginary part.
For our integral, the imaginary part should be exactly zero (which is true to the machine accuracy here) so it can be ignored.

# Q2.) What answer did SciPy give? Did it agree with your calculation? How did it compare with the Trapezoidal rule calculations with $N=10$ and $N=100$?

SciPy agreed with our exact result up to the 13th place after the decimal point (taking rounding into account).
This was much better that either of the Trapezoidal rule calculations.


Here's an integral that the Trapezoidal method will struggle with:
$$
\int_{0.00001}^1\,\dfrac{dx}{x}\,.
$$

Try it using using our Trapezoidal method (varying the number of steps) and with the ***integrate.quad*** function from SciPy and compare with the exact answer in the cell below.  (Note that if you use too many steps in the Trapezoidal method it begins to run noticeably slow.)

In [24]:
def f(x):
    return 1/x

print("Trapezoidal with N=1000 gives      ",Trapezoidal(f,0.00001,1,1000))
print("Trapezoidal with N=10000 gives     ",Trapezoidal(f,0.00001,1,10000))
print("Trapezoidal with N=100000 gives    ",Trapezoidal(f,0.00001,1,100000))
print("Trapezoidal with N=1000000 gives   ",Trapezoidal(f,0.00001,1,1000000))
print("integrate.quad gives              ",integrate.quad(f,0.00001,1))
print("exact answer is.                   ",np.log(1/.00001))

Trapezoidal with N=1000 gives       57.46815049144616
Trapezoidal with N=10000 gives      14.634053878291443
Trapezoidal with N=100000 gives     11.514622673621659
Trapezoidal with N=1000000 gives    11.512933798055156
integrate.quad gives               (11.51292546497023, 7.545599548893293e-11)
exact answer is.                    11.512925464970229


# Q3.) What value of $N$ was needed for the Trapezoidal calculation to get agreement with the exact answer to 2 places after the decimal point?  How did did SciPy do on this integral?

For Trapezoidal calculation needed about a million steps (actually around N=7x10^5).  (If you go over a million steps the calculation starts to become noticeably slower.) SciPy hit it right on to the last decimal point.

# Let's go back to using our Trapezoidal rule code and see how it can be applied for analyzing physical situations.

Assume we have 3 rods with mass densities of:

- $\rho_A = 5x\ \dfrac{kg}{m}$
- $\rho_B = 3x(20-x)\ \dfrac{kg}{m}$
- $\rho_C = 3x^2(6-x) \ \dfrac{kg}{m}$. 

They have lengths of:

- $l_A = 4m$ 
- $l_B = 10m$
- $l_C = 6m$

In the cells below, use our Trapezoidal function to calculate each of their masses. We have given you an outline for object A.

In [9]:
# Find mass of A here

# Density function for A
def rho_A(x):
    return 5*x# Put density function of A here

# Utilize Trapezoidal method here from 0 to 4 (length of rod)
mass_A = Trapezoidal(rho_A, 0, 4, 1000)
print("Mass of A is: ", mass_A)

Mass of A is:  39.99999999999999


In [10]:
# Find mass of B here

# Density function for B
def rho_B(x):
    return 3*x*(20-x)# Put density function of B here

# Utilize Trapezoidal method here from 0 to 10 (length of rod)
mass_B = Trapezoidal(rho_B, 0, 10, 1000)
print("Mass of B is: ", mass_B)

Mass of B is:  1999.999500000001


In [11]:
# Find mass of C here

# Density function for C
def rho_C(x):
    return 3*x**2*(6-x)# Put density function of C here

# Utilize Trapezoidal method here from 0 to 6 (length of rod)
mass_C = Trapezoidal(rho_C, 0, 6, 1000)
print("Mass of C is: ", mass_C)

Mass of C is:  323.99967599999985


# Q4.) What are the masses of our three rods?

mass_A=40 kg, 
mass_B=2000 kg, 
mass_C=324 kg

Now using the three values calculated above, let's calculate the center of mass of each rod. Recall that to find the center of mass for a 1D object: 

$\bar{x} = \dfrac{\int_a^b x\rho(x) dx}{m}$, where $m$ is the mass of the object.

You already have $m$, but now you need to find the integration of the top part. Let us define $g(x) = x\rho(x)$.

In the cells below, find the center of the mass using the Trapezoidal method. We have given you an outline of how to find $\bar{x}_A$.

In [25]:
# Find center of mass of A here

def g_A(x):
    return x*rho_A(x)

# Utilize Trapezoidal method here from 0 to 4 (length of rod). Then divide by mass of A
center_A = Trapezoidal(g_A, 0, 4, 1000)/mass_A
center_A

2.6666680000000014

In [13]:
# Find center of mass of B here

def g_B(x):
    return x*rho_B(x)

# Utilize Trapezoidal method here from 0 to 10 (length of rod). Then divide by mass of B
center_B = Trapezoidal(g_B, 0, 10, 1000)/mass_B
center_B

6.250002812500697

In [14]:
# Find center of mass of C here

def g_C(x):
    return x*rho_C(x)

# Utilize Trapezoidal method here from 0 to 6 (length of rod). Then divide by mass of C
center_C = Trapezoidal(g_C, 0, 6, 1000)/mass_C
center_C

3.599997600000001

# Q5.) What are the center of masses of our three rods?

The center of masses are:
center_A=2.67 m, 
center B=6.25 m, 
center C=3.60 m

# Notebook Wrap-up. 
Run the cell below and copy-paste your answers into their corresponding cells.

In [15]:
from IPython.display import HTML
HTML(
"""
<iframe 
	src="https://forms.gle/cSUgKLqC8Qo2Ph838" 
	width="100%" 
	height="1200px" 
	frameborder="0" 
	marginheight="0" 
	marginwidth="0">
	Loading...
</iframe>
"""
)

# Congrulations! Another week completed!

Don't be shy to practice things you are having a bit of a hard time with, or maybe you are getting everything that's being taught. Either way, it's really your call what you want to do now: relax, study, drink tea, etc. [To help lighten the mood during midterms, here is picture that demonstrates the founding of Physics.](https://claesjohnsonmathscience.files.wordpress.com/2011/12/newton2.jpg)